File: src/main/java/com/danielremsburg/jaffolding/bridge/ComponentRegistry.java
package com.danielremsburg.jaffolding.bridge;

import org.teavm.jso.JSBody;
import org.teavm.jso.JSObject;
import org.teavm.jso.JSProperty;

/**
 * Registry for mapping between Java and JavaScript components.
 * This allows components to be shared between the two environments.
 */
public class ComponentRegistry {
    
    /**
     * Registers a Java component implementation.
     * @param id The component identifier
     * @param className The fully qualified Java class name
     */
    public static void registerJavaComponent(String id, String className) {
        registerJavaComponentImpl(id, className, "{}");
    }
    
    /**
     * Registers a Java component implementation with options.
     * @param id The component identifier
     * @param className The fully qualified Java class name
     * @param options The registration options as a JSON string
     */
    public static void registerJavaComponent(String id, String className, String options) {
        registerJavaComponentImpl(id, className, options);
    }
    
    /**
     * Gets a component implementation.
     * @param id The component identifier
     * @return The component implementation info as a JSObject
     */
    public static JSObject getComponent(String id) {
        return getComponentImpl(id, null);
    }
    
    /**
     * Gets a component implementation with a preferred type.
     * @param id The component identifier
     * @param preferredType The preferred implementation type ('js' or 'java')
     * @return The component implementation info as a JSObject
     */
    public static JSObject getComponent(String id, String preferredType) {
        return getComponentImpl(id, preferredType);
    }
    
    /**
     * Creates a component instance.
     * @param id The component identifier
     * @param args The constructor arguments as a JSObject array
     * @return The component instance as a JSObject
     */
    public static JSObject createComponent(String id, JSObject[] args) {
        return createComponentImpl(id, args, "{}");
    }
    
    /**
     * Creates a component instance with options.
     * @param id The component identifier
     * @param args The constructor arguments as a JSObject array
     * @param options The creation options as a JSON string
     * @return The component instance as a JSObject
     */
    public static JSObject createComponent(String id, JSObject[] args, String options) {
        return createComponentImpl(id, args, options);
    }
    
    /**
     * Enables or disables the bridge.
     * @param enabled Whether the bridge is enabled
     */
    public static void setBridgeEnabled(boolean enabled) {
        setBridgeEnabledImpl(enabled);
    }
    
    /**
     * Checks if the bridge is enabled.
     * @return Whether the bridge is enabled
     */
    public static boolean isBridgeEnabled() {
        return isBridgeEnabledImpl();
    }
    
    /**
     * Checks if a component is registered.
     * @param id The component identifier
     * @return Whether the component is registered
     */
    public static boolean isComponentRegistered(String id) {
        return isComponentRegisteredImpl(id);
    }
    
    /**
     * Checks if a JavaScript component is available.
     * @param id The component identifier
     * @return Whether the JavaScript component is available
     */
    public static boolean isJSComponentAvailable(String id) {
        return isJSComponentAvailableImpl(id);
    }
    
    /**
     * Checks if a Java component is available.
     * @param id The component identifier
     * @return Whether the Java component is available
     */
    public static boolean isJavaComponentAvailable(String id) {
        return isJavaComponentAvailableImpl(id);
    }
    
    // Native implementations
    
    @JSBody(params = {"id", "className", "options"}, script = 
        "try {" +
        "  if (window.jaffolding && window.jaffolding.ComponentRegistry) {" +
        "    window.jaffolding.ComponentRegistry.getInstance().registerJavaComponent(id, className, JSON.parse(options));" +
        "    return true;" +
        "  }" +
        "  return false;" +
        "} catch (e) {" +
        "  console.error('Error registering Java component:', e);" +
        "  return false;" +
        "}")
    private static native boolean registerJavaComponentImpl(String id, String className, String options);
    
    @JSBody(params = {"id", "preferredType"}, script = 
        "try {" +
        "  if (window.jaffolding && window.jaffolding.ComponentRegistry) {" +
        "    return window.jaffolding.ComponentRegistry.getInstance().getComponent(id, preferredType);" +
        "  }" +
        "  return null;" +
        "} catch (e) {" +
        "  console.error('Error getting component:', e);" +
        "  return null;" +
        "}")
    private static native JSObject getComponentImpl(String id, String preferredType);
    
    @JSBody(params = {"id", "args", "options"}, script = 
        "try {" +
        "  if (window.jaffolding && window.jaffolding.ComponentRegistry) {" +
        "    return window.jaffolding.ComponentRegistry.getInstance().createComponent(id, args, JSON.parse(options));" +
        "  }" +
        "  return null;" +
        "} catch (e) {" +
        "  console.error('Error creating component:', e);" +
        "  return null;" +
        "}")
    private static native JSObject createComponentImpl(String id, JSObject[] args, String options);
    
    @JSBody(params = {"enabled"}, script = 
        "try {" +
        "  if (window.jaffolding && window.jaffolding.ComponentRegistry) {" +
        "    window.jaffolding.ComponentRegistry.getInstance().setBridgeEnabled(enabled);" +
        "    return true;" +
        "  }" +
        "  return false;" +
        "} catch (e) {" +
        "  console.error('Error setting bridge enabled:', e);" +
        "  return false;" +
        "}")
    private static native boolean setBridgeEnabledImpl(boolean enabled);
    
    @JSBody(params = {}, script = 
        "try {" +
        "  if (window.jaffolding && window.jaffolding.ComponentRegistry) {" +
        "    return window.jaffolding.ComponentRegistry.getInstance().isBridgeEnabled();" +
        "  }" +
        "  return false;" +
        "} catch (e) {" +
        "  console.error('Error checking if bridge is enabled:', e);" +
        "  return false;" +
        "}")
    private static native boolean isBridgeEnabledImpl();
    
    @JSBody(params = {"id"}, script = 
        "try {" +
        "  if (window.jaffolding && window.jaffolding.ComponentRegistry) {" +
        "    return window.jaffolding.ComponentRegistry.getInstance().getComponent(id) !== null;" +
        "  }" +
        "  return false;" +
        "} catch (e) {" +
        "  console.error('Error checking if component is registered:', e);" +
        "  return false;" +
        "}")
    private static native boolean isComponentRegisteredImpl(String id);
    
    @JSBody(params = {"id"}, script = 
        "try {" +
        "  if (window.jaffolding && window.jaffolding.ComponentRegistry) {" +
        "    const component = window.jaffolding.ComponentRegistry.getInstance().getComponent(id, 'js');" +
        "    return component !== null && component.type === 'js';" +
        "  }" +
        "  return false;" +
        "} catch (e) {" +
        "  console.error('Error checking if JS component is available:', e);" +
        "  return false;" +
        "}")
    private static native boolean isJSComponentAvailableImpl(String id);
    
    @JSBody(params = {"id"}, script = 
        "try {" +
        "  if (window.jaffolding && window.jaffolding.ComponentRegistry) {" +
        "    const component = window.jaffolding.ComponentRegistry.getInstance().getComponent(id, 'java');" +
        "    return component !== null && component.type === 'java';" +
        "  }" +
        "  return false;" +
        "} catch (e) {" +
        "  console.error('Error checking if Java component is available:', e);" +
        "  return false;" +
        "}")
    private static native boolean isJavaComponentAvailableImpl(String id);
}

File: src/main/resources/static/js/jaffolding/bridge/ComponentRegistry.js
/**
 * Registry for mapping between Java and JavaScript components.
 * This allows components to be shared between the two environments.
 */
(function() {
  class ComponentRegistry {
    constructor() {
      this.components = {};
      this.javaComponents = {};
      this.jsComponents = {};
      this.bridgeEnabled = true;
    }

    /**
     * Registers a JavaScript component implementation.
     * @param {string} id - The component identifier
     * @param {Function} constructor - The component constructor
     * @param {Object} [options] - Registration options
     */
    registerJSComponent(id, constructor, options = {}) {
      this.jsComponents[id] = {
        constructor,
        options
      };
      
      // Also register in the general registry
      this.components[id] = {
        type: 'js',
        implementation: constructor,
        options
      };
      
      console.log(`Registered JS component: ${id}`);
    }

    /**
     * Registers a Java component implementation.
     * @param {string} id - The component identifier
     * @param {string} className - The fully qualified Java class name
     * @param {Object} [options] - Registration options
     */
    registerJavaComponent(id, className, options = {}) {
      this.javaComponents[id] = {
        className,
        options
      };
      
      // Also register in the general registry
      this.components[id] = {
        type: 'java',
        implementation: className,
        options
      };
      
      console.log(`Registered Java component: ${id}`);
    }

    /**
     * Gets a component implementation.
     * @param {string} id - The component identifier
     * @param {string} [preferredType] - The preferred implementation type ('js' or 'java')
     * @returns {Object} - The component implementation info
     */
    getComponent(id, preferredType) {
      const component = this.components[id];
      if (!component) return null;
      
      // If no preference or the preferred type matches, return the component
      if (!preferredType || component.type === preferredType) {
        return component;
      }
      
      // If the preferred type is different, check if the other type is available
      const otherType = preferredType === 'js' ? 'java' : 'js';
      const otherComponent = this.components[id + '_' + otherType];
      
      // Return the preferred type if available, otherwise the original
      return otherComponent || component;
    }

    /**
     * Creates a component instance.
     * @param {string} id - The component identifier
     * @param {Array} args - The constructor arguments
     * @param {Object} [options] - Creation options
     * @returns {Object} - The component instance
     */
    createComponent(id, args = [], options = {}) {
      const preferredType = options.preferredType || (window.$rt_javaClassesLoaded ? 'java' : 'js');
      const component = this.getComponent(id, preferredType);
      
      if (!component) {
        console.error(`Component not found: ${id}`);
        return null;
      }
      
      try {
        if (component.type === 'js') {
          // Create JavaScript component
          return new component.implementation(...args);
        } else {
          // Create Java component
          if (!window.$rt_javaClassesLoaded) {
            console.warn(`Cannot create Java component ${id}: TeaVM not loaded`);
            
            // Try to fall back to JS implementation
            const jsComponent = this.getComponent(id, 'js');
            if (jsComponent) {
              console.log(`Falling back to JS implementation for ${id}`);
              return new jsComponent.implementation(...args);
            }
            
            return null;
          }
          
          return window.jaffolding.JavaBridge.createComponent(component.implementation, args);
        }
      } catch (e) {
        console.error(`Error creating component ${id}:`, e);
        return null;
      }
    }

    /**
     * Enables or disables the bridge.
     * @param {boolean} enabled - Whether the bridge is enabled
     */
    setBridgeEnabled(enabled) {
      this.bridgeEnabled = enabled;
    }

    /**
     * Checks if the bridge is enabled.
     * @returns {boolean} - Whether the bridge is enabled
     */
    isBridgeEnabled() {
      return this.bridgeEnabled;
    }

    /**
     * Gets the singleton instance.
     * @returns {ComponentRegistry} - The singleton instance
     */
    static getInstance() {
      if (!ComponentRegistry.instance) {
        ComponentRegistry.instance = new ComponentRegistry();
      }
      return ComponentRegistry.instance;
    }
  }

  // Create the singleton instance
  ComponentRegistry.instance = null;

  // Export the ComponentRegistry class
  window.jaffolding = window.jaffolding || {};
  window.jaffolding.ComponentRegistry = ComponentRegistry;
  
  // Register core components
  const registry = ComponentRegistry.getInstance();
  
  // Register KDEDesktop
  if (window.jaffolding.KDEDesktop) {
    registry.registerJSComponent('kde-desktop', window.jaffolding.KDEDesktop);
  }
  
  // Register AppManager
  if (window.jaffolding.AppManager) {
    registry.registerJSComponent('app-manager', window.jaffolding.AppManager);
  }
  
  // Register BrowserWindow
  if (window.jaffolding.BrowserWindow) {
    registry.registerJSComponent('browser-window', window.jaffolding.BrowserWindow);
  }
  
  // Register Calculator
  if (window.jaffolding.Calculator) {
    registry.registerJSComponent('calculator', window.jaffolding.Calculator);
  }
  
  console.log('Component Registry initialized');
})();